geopandas/io/tests/test_file.py

```diff
@@ -139,15 +140,37 @@ def test_to_file_bool(tmpdir, driver, ext):
     assert_correct_driver(tempfilename, ext)
 
 
-def test_to_file_datetime(tmpdir):
+dt_resolution = ["ms", "s", "min", "h", "D"]
+
+
+@pytest.mark.parametrize("dt_res", dt_resolution)
+@pytest.mark.parametrize("driver,ext", driver_ext_pairs)
+def test_to_file_datetime(tmpdir, driver, ext, dt_res):
     """Test writing a data file with the datetime column type"""
-    tempfilename = os.path.join(str(tmpdir), "test_datetime.gpkg")
+    if ext in (".shp", ""):
+        pytest.skip(f"Driver corresponding to ext {ext} doesn't support dt fields")
+    tempfilename = os.path.join(str(tmpdir), f"test_datetime{ext}")
     point = Point(0, 0)
-    now = datetime.datetime.now()
+    now = datetime.datetime(2021, 11, 21, 1, 7, 43, 17500)
+    # left: 1637456996510000000, right: 1637456996509998000
+    # left: 1637456863175000000, right: 1637456863174999000
+    # left: 1637457037286000000, right: 1637457037285999000
+
     df = GeoDataFrame({"a": [1, 2], "b": [now, now]}, geometry=[point, point], crs=4326)
-    df.to_file(tempfilename, driver="GPKG")
+    if compat.FIONA_GE_1814:
+        fiona_precision_limit = "ms"
+    else:
+        fiona_precision_limit = "s"
+    df["b"] = df["b"].dt.round(freq=fiona_precision_limit)
+
+    if dt_res is not None:
+        df["b"] = df["b"].dt.round(freq=dt_res)
+    df.to_file(tempfilename, driver=driver)
     df_read = read_file(tempfilename)
-    assert_geoseries_equal(df.geometry, df_read.geometry)
+
+    assert_geodataframe_equal(df.drop(columns=["b"]), df_read.drop(columns=["b"]))
+    # fiona seems to introduce up to 2 microseconds of error on reading back in
+    assert_allclose(df["b"].view("int64"), df["b"].view("int64"), atol=2000)
```

Comment:
This is strange .. Should we consider that a bug in Fiona?

Reply:
Looking a bit into this, so I think GDAL itself only supports milliseconds? (you might have noted this elsewhere)

Patching a local dev install of fiona with some prints:

```
In [21]: df = geopandas.GeoDataFrame({'col': pd.date_range("2012-01-01 09:01:01.123456", periods=1), "geometry": geopandas.points_from_xy([0], [1])})

In [22]: df.to_file("test_datetime.gpkg")
seconds: 1 - microseconds: 123456
seconds: 1.123456

In [23]: geopandas.read_file("test_datetime.gpkg")
Getting date (float seconds, fractional, integer):  1.1230000257492065 0.12300002574920654 1.0
seconds: 1 - microseconds: 123000
Out[23]: 
                          col                 geometry
0  2012-01-01T09:01:01.123000  POINT (0.00000 1.00000)
1  2012-01-02T09:01:01.123000  POINT (1.00000 2.00000)
```

So in this case I actually don't get wrong microseconds in the end result (but that might be value specific).

Now, the value that fiona gets from GDAL does have some additional values: 1.1230000257492065
But this seems to be because of storing 1.123 as float32, I think:

```
In [29]: np.array([1.123], dtype="float32")
Out[29]: array([1.123], dtype=float32)

In [30]: np.array([1.123], dtype="float32")[0].item()
Out[30]: 1.1230000257492065
```

Reply:
Indeed, using the datetime value from the test above, I see the 2 microseconds off:

```
In [32]: now = datetime.datetime(2021, 11, 21, 1, 7, 43, 17500)

In [33]: df = geopandas.GeoDataFrame({'col': [now], "geometry": geopandas.points_from_xy([0], [1])})

In [34]: df
Out[34]: 
                         col                 geometry
0 2021-11-21 01:07:43.017500  POINT (0.00000 1.00000)

In [35]: df.to_file("test_datetime.gpkg")
seconds: 43 - microseconds: 17500
seconds: 43.0175

In [36]: geopandas.read_file("test_datetime.gpkg")
Getting date (float seconds, fractional, integer):  43.018001556396484 0.018001556396484375 43.0
seconds: 43 - microseconds: 18002
Out[36]: 
                          col                 geometry
0  2021-11-21T01:07:43.018002  POINT (0.00000 1.00000)
```

I think we can probably "fix" this on our side to remove any microseconds, knowing that those can never come from GDAL (and I think fiona should do the same)

Reply:
> Should we consider that a bug in Fiona?

I would think so, but I admittedly don't know much about fiona.

What you've said about the error on reading and fiona only supporting up to miliseconds (https://github.com/Toblerity/Fiona/issues/744) is consistent with what I've found. Nice to know it's from float rounding, given that explains what is happening, dropping digits past miliseconds seems logical.